### ISSUE TITLE
Add support for Vencrypt x509 security type

### ISSUE THREAD

**Issue by Kiisu-Master** [gujjwal00/avnc#249]
### Info:
Android 11
AVNC 2.6.1 from F-Droid

### Issue
Wanted to connect to a raspberry pi 4 from my phone, but it always says "Disconnected". Connecting from my laptop with TigerVNC client works fine, and connecting with bVNC Free from F-Droid on the same phone also works. I tried both hostname and IP address. Could this be a compatibility issue between the server and client? The discovery feature also didn't find the raspberry pi after waiting for a while.

**Comment by gujjwal00** [comment]
Please go to AVNC Settings => Tools => Logs. Copy the logs, and pate here.

**Comment by Kiisu-Master** [comment]
Other VNC clients usually ask something about if I trust the connection or credential stuff but this one never did. Maybe that's related. I cut the log to the part where the connection is attempted.

````python
09-06 20:07:26.492 30461 30542 I NativeVnc: Received protocol version 3.8
09-06 20:07:26.492 30461 30542 I NativeVnc: VNC server supports protocol version 3.8 (viewer 3.8)
09-06 20:07:26.493 30461 30461 I InputTransport: Create ARC handle: 0xd1b88d00
09-06 20:07:26.495 30461 30542 I NativeVnc: We have 3 security types to read
09-06 20:07:26.495 30461 30542 I NativeVnc: 0) Received security type 19
09-06 20:07:26.495 30461 30542 I NativeVnc: Selecting security type 19 (0/3 in the list)
09-06 20:07:26.495 30461 30542 I NativeVnc: 1) Received security type 129
09-06 20:07:26.496 30461 30542 I NativeVnc: 2) Received security type 5
09-06 20:07:26.496 30461 30542 I NativeVnc: Selected Security Scheme 19
09-06 20:07:26.498 30461 30542 I NativeVnc: Got VeNCrypt version 0.2 from server.
09-06 20:07:26.501 30461 30542 I NativeVnc: We have 1 security types to read
09-06 20:07:26.501 30461 30542 I NativeVnc: 0) Received security type 262
09-06 20:07:26.501 30461 30542 I NativeVnc: Selecting security type 262
09-06 20:07:26.504 30461 30542 I NativeVnc: OpenSSL version wolfSSL 5.7.2 initialized.
09-06 20:07:26.504 30461 30542 E NativeVnc: Unsupported credential type requested
09-06 20:07:26.504 30461 30542 I NativeVnc: Reading credential failed
09-06 20:07:26.505 30461 30542 E NativeVnc: errnoToStr: (11 Try again)
09-06 20:07:26.505 30461 30542 E VncViewModel: Connection failed
09-06 20:07:26.505 30461 30542 E VncViewModel: java.io.IOException: 
09-06 20:07:26.505 30461 30542 E VncViewModel: 	at com.gaurav.avnc.vnc.VncClient.connect(VncClient.kt:25)
09-06 20:07:26.505 30461 30542 E VncViewModel: 	at com.gaurav.avnc.viewmodel.VncViewModel.access$connect(VncViewModel.kt:221)
09-06 20:07:26.505 30461 30542 E VncViewModel: 	at com.gaurav.avnc.viewmodel.VncViewModel$launchConnection$1.invoke(VncViewModel.kt:6)
09-06 20:07:26.505 30461 30542 E VncViewModel: 	at kotlin.concurrent.ThreadsKt$thread$thread$1.run(Thread.kt:3)

````

**Comment by gujjwal00** [comment]
So the issue is that VNC server is using security type 262, which is not enabled/supported in AVNC. Couple of questions:
- Does TigerVNC client also asks if server is trusted before connecting?
- Did you configure a x509 certificate when setting up your server, or are you using default settings (so a certificate was probably generated automatically)? 

As for a workaround, you might be able to change the security type in your VNC server settings. If you don't know how, please let me know which VNC server you are running on your Pi.

**Comment by Kiisu-Master** [comment]
I followed the documentation on raspberry pi website. In TigerVNC it tells me it's connected to WayVNC.

Also the [documentation](https://www.raspberrypi.com/documentation/computers/remote-access.html#connect-to-a-vnc-server) includes the screenshots of TigerVNC warning messages which are to be ignored. I got a similar warning message in the other VNC client on phone.

**Comment by gujjwal00** [comment]
For WayVNC, you should be able to allow other security types:

1. Connect to Raspberry Pi from TigerVNC viewer or bVNC.
2. Locate the WayVNC config file. It should be at `$HOME/.config/wayvnc/config`.
3. Open the file in text editor
4. Add a new line with the following text: `relax_encryption=true`
5. Save file & restart VNC server
7. Try to connect using AVNC

**Comment by gujjwal00** [comment]
If that doesn't work, pleas tell me the WayVNC version installed on your PI.

**Comment by Kiisu-Master** [comment]
Wayvnc version: 0.9.0~dev-1
Used the setting for relaxed encryption but still not connecting.
Error in the logs seems to be the same.

**Comment by Kiisu-Master** [comment]
Actually logs mention another security type that wasn't available before: type 30. It still uses 19 though.

**Comment by gujjwal00** [comment]
So changes to config file worked, that's good. But as you said, LibVNC is still selecting the security type 19. To fix this, I have added an option to force security type 30.

To test this, please install the following APK. After adding the server to this test APK, open advanced server options (Long-press on server => Edit => Advanced). Change `Security` to `macOS` and save it . AVNC should select the correct security type now.  
Test APK: https://github.com/gujjwal00/avnc/actions/runs/10993552270/artifacts/1965921601

**Comment by gujjwal00** [comment]
@curable-online The issue `no shared cipher` is most probably caused by ADH algorithms not being enabled in x11vnc. This can happen if x11vnc is using newer OpenSSL versions, which if I recall correctly, require few more tweaks to enable these legacy ciphers. BTW, I tried TigerVNC viewer and it failed with the same error for me.

That said, I would encourage you to use SSH if you can, because it has better protection against MITM attacks.

**Comment by gujjwal00** [comment]
Error `routines::unexpected eof while reading` is expected when VeNCrypt sub-type: 261 (rfbVencryptX509Vnc) is used. 

When you disable AnonTLS in x11vnc, it won't advertise those sub-types to clients, and clients will end up choosing `261`. Problem with AVNC is that it does not support X509 sub-types of VeNCrypt. But LibVNCClient, VNC library used by AVNC,  has no way to tell it that AVNC doesn't support these, without disabling other sub-types too. So it initially selects the sub-type `261`, but fails during TLS handshake because AVNC cannot provide required X509 certificates.

For X509 sub-types, LibVNCClient right now expects a CA certificate to validate server's certificate. It has no option to work like TigerVNC where you can use self-signed certificates on server, and viewer will warn you before connecting to new servers.

**Comment by gujjwal00** [comment]
> Does this work with AVNC if I try?

No, I don't believe so. LibVNCClient uses OpenSSL (wolfSSL in case of AVNC), which I believe have no idea about the Android's CA certificates. Android's CA certificate are primarily intended to be used by connections initiated from Java, or by apps with specific support for them.

**Comment by Kiisu-Master** [comment]
I tried the test APK you made with macOS security and it works now! Its way smoother connection than any other VNC client. How is that possible?

**Comment by gujjwal00** [comment]
> How is that possible?

A lot of credit goes to libjpeg-turbo library, and further improvements are gained through OpenGL rendering.  

BTW, v2.6.2 is now available with this security option.

**Comment by k2helix** [comment]
Where can I enable that option? I cannot find it.

**Comment by gujjwal00** [comment]
@k2helix please post the logs from AVNC settings after you see the "Authentication failed" error.

**Comment by k2helix** [comment]
Sure, here they are:

```
10-29 05:46:27.235 24652 24752 I NativeVnc: Received protocol version 3.8
10-29 05:46:27.235 24652 24752 I NativeVnc: VNC server supports protocol version 3.8 (viewer 3.8)
10-29 05:46:27.238 24652 24752 I NativeVnc: We have 3 security types to read
10-29 05:46:27.238 24652 24752 I NativeVnc: 0) Received security type 19
10-29 05:46:27.238 24652 24752 I NativeVnc: 1) Received security type 129
10-29 05:46:27.238 24652 24752 I NativeVnc: 2) Received security type 5
10-29 05:46:27.238 24652 24752 E NativeVnc: Unknown authentication scheme from VNC server: 19, 129, 5
10-29 05:46:27.240 24652 24665 W com.gaurav.avnc: Cleared Reference was only reachable from finalizer (only reported once)
10-29 05:46:27.241 24652 24752 E VncViewModel: Connection failed
10-29 05:46:27.241 24652 24752 E VncViewModel: java.io.IOException: Authentication failed
10-29 05:46:27.241 24652 24752 E VncViewModel: 	at com.gaurav.avnc.vnc.VncClient.connect(VncClient.kt:116)
10-29 05:46:27.241 24652 24752 E VncViewModel: 	at com.gaurav.avnc.viewmodel.VncViewModel.access$connect(VncViewModel.kt:221)
10-29 05:46:27.241 24652 24752 E VncViewModel: 	at com.gaurav.avnc.viewmodel.VncViewModel$launchConnection$1.invoke(VncViewModel.kt:6)
10-29 05:46:27.241 24652 24752 E VncViewModel: 	at kotlin.concurrent.ThreadsKt$thread$thread$1.run(Thread.kt:3)
10-29 05:46:27.603 24652 24652 D VRI[VncActivity]: visibilityChanged oldVisibility=true newVisibility=false
10-29 05:46:27.611 24652 24673 D HWUI    : endAllActiveAnimators on 0x300c9bd6af31c00 (RippleDrawable) with handle 0x300c88aeb0d3630
10-29 05:46:27.615 24652 24652 W WindowOnBackDispatcher: sendCancelIfRunning: isInProgress=false callback=android.view.ViewRootImpl$$ExternalSyntheticLambda11@14dbb7e
10-29 05:46:28.118 24652 24652 D VRI[VncActivity]: visibilityChanged oldVisibility=true newVisibility=false
10-29 05:46:28.133 24652 24652 W WindowOnBackDispatcher: sendCancelIfRunning: isInProgress=false callback=android.view.ViewRootImpl$$ExternalSyntheticLambda11@b3925e6
10-29 05:46:29.962 24652 24652 D VRI[StartupActivity]: visibilityChanged oldVisibility=true newVisibility=false
10-29 05:46:29.971 24652 24673 D HWUI    : endAllActiveAnimators on 0x200c9bd6ae1dc00 (NavigationMenuItemView) with handle 0xa00c88aeb1e7cf0
10-29 05:46:30.469 24652 24652 W WindowOnBackDispatcher: OnBackInvokedCallback is not enabled for the application.
10-29 05:46:30.469 24652 24652 W WindowOnBackDispatcher: Set 'android:enableOnBackInvokedCallback="true"' in the application manifest.
10-29 05:46:34.480 24759 24759 W libc    : Access denied finding property "ro.debuggable"
10-29 05:46:34.473 24759 24759 W logcat  : type=1400 audit(0.0:63313): avc:  denied  { read } for  name="u:object_r:userdebug_or_eng_prop:s0" dev="tmpfs" ino=448 scontext=u:r:untrusted_app:s0:c27,c257,c512,c768 tcontext=u:object_r:userdebug_or_eng_prop:s0 tclass=file permissive=0 app=com.gaurav.avnc
```
I selected the "macOS" security under advanced settings. The other options either throw the same error or just don't work as if it was set to Auto.

**Comment by gujjwal00** [comment]
Please update the server config as described above: https://github.com/gujjwal00/avnc/issues/249#issuecomment-2336399110

**Comment by k2helix** [comment]
Aaand that works. Thanks for helping me out. I guess this is the whole point of the issue and that you have already explained it, but it seems to me that this relax_encryption is a workaround instead of a direct solution. Wayvnc closed issues related to this because "Security trumps compatibility"

**Comment by Raptor-TK** [comment]
I also have this problem but when trying to connect to a PiKVM. 
Currently bVNC seems to connect fine but same issue with AVNC.

I can't try the test APK as the link gives a 404 error.

**Logs:**
10-30 18:03:14.130 26986 29448 I NativeVnc: Received protocol version 3.8
10-30 18:03:14.130 26986 29448 I NativeVnc: VNC server supports protocol version 3.8 (viewer 3.8)
10-30 18:03:14.137 26986 26986 D SurfaceView@e6311f9: 241373689 updateSurface: has no frame
10-30 18:03:14.145 26986 26986 D SurfaceView@e6311f9: 241373689 updateSurface: has no frame
10-30 18:03:14.153 26986 26986 D SurfaceView@e6311f9: 241373689 updateSurface: has no frame
10-30 18:03:14.162 26986 26986 D SurfaceView@e6311f9: 241373689 updateSurface: has no frame
10-30 18:03:14.170 26986 26986 D SurfaceView@e6311f9: 241373689 updateSurface: has no frame
10-30 18:03:14.178 26986 26986 D SurfaceView@e6311f9: 241373689 updateSurface: has no frame
10-30 18:03:14.187 26986 26986 D SurfaceView@e6311f9: 241373689 updateSurface: has no frame
10-30 18:03:14.195 26986 26986 D SurfaceView@e6311f9: 241373689 updateSurface: has no frame
10-30 18:03:14.204 26986 26986 D SurfaceView@e6311f9: 241373689 updateSurface: has no frame
10-30 18:03:14.235 26986 29448 I NativeVnc: We have 1 security types to read
10-30 18:03:14.235 26986 29448 I NativeVnc: 0) Received security type 19
10-30 18:03:14.235 26986 29448 I NativeVnc: Selecting security type 19 (0/1 in the list)
10-30 18:03:14.235 26986 29448 I NativeVnc: Selected Security Scheme 19
10-30 18:03:14.326 26986 29448 I NativeVnc: Got VeNCrypt version 0.2 from server.
10-30 18:03:14.336 26986 26986 I ViewRootImpl@5e2e7f[StartupActivity]: handleAppVisibility mAppVisible = true visible = false
10-30 18:03:14.346 26986 26986 I ViewRootImpl@5e2e7f[StartupActivity]: destroyHardwareResources: Callers=android.view.ViewRootImpl.performTraversals:3993 android.view.ViewRootImpl.doTraversal:3345 android.view.ViewRootImpl$TraversalRunnable.run:11441 android.view.Choreographer$CallbackRecord.run:1690 android.view.Choreographer$CallbackRecord.run:1699 android.view.Choreographer.doCallbacks:1154 android.view.Choreographer.doFrame:1080 android.view.Choreographer$FrameDisplayEventReceiver.run:1647 android.os.Handler.handleCallback:958 android.os.Handler.dispatchMessage:99 
10-30 18:03:14.361 26986 26986 I ViewRootImpl@5e2e7f[StartupActivity]: Relayout returned: old=(0,0,968,2376) new=(0,0,968,2376) relayoutAsync=false req=(968,2376)8 dur=5 res=0x402 s={false 0x0} ch=true seqId=0
10-30 18:03:14.361 26986 26986 I ViewRootImpl@5e2e7f[StartupActivity]: stopped(true) old = false
10-30 18:03:14.361 26986 26986 D ViewRootImpl@5e2e7f[StartupActivity]: WindowStopped on com.gaurav.avnc/com.gaurav.avnc.StartupActivity set to true
10-30 18:03:14.375 26986 26986 I ViewRootImpl@5e2e7f[StartupActivity]: Relayout returned: old=(0,0,968,2376) new=(0,0,968,2376) relayoutAsync=false req=(968,2376)8 dur=5 res=0x402 s={false 0x0} ch=false seqId=0
10-30 18:03:14.430 26986 29448 I NativeVnc: We have 3 security types to read
10-30 18:03:14.430 26986 29448 I NativeVnc: 0) Received security type 256
10-30 18:03:14.431 26986 29448 I NativeVnc: 1) Received security type 262
10-30 18:03:14.431 26986 29448 I NativeVnc: 2) Received security type 259
10-30 18:03:14.431 26986 29448 I NativeVnc: Selecting security type 262
10-30 18:03:14.527 26986 29448 E NativeVnc: Unsupported credential type requested
10-30 18:03:14.528 26986 29448 I NativeVnc: Reading credential failed
10-30 18:03:14.528 26986 29448 E NativeVnc: errnoToStr: (11 Try again)
10-30 18:03:14.528 26986 29448 E VncViewModel: Connection failed
10-30 18:03:14.528 26986 29448 E VncViewModel: java.io.IOException: 
10-30 18:03:14.528 26986 29448 E VncViewModel: 	at com.gaurav.avnc.vnc.VncClient.connect(VncClient.kt:116)
10-30 18:03:14.528 26986 29448 E VncViewModel: 	at com.gaurav.avnc.viewmodel.VncViewModel.access$connect(VncViewModel.kt:221)
10-30 18:03:14.528 26986 29448 E VncViewModel: 	at com.gaurav.avnc.viewmodel.VncViewModel$launchConnection$1.invoke(VncViewModel.kt:6)

**Comment by gujjwal00** [comment]
Let me summarize the situation to make it easier to follow:

### The Problem
During VNC connection setup, both client & server has to agree upon a security type. By default, none of the security types offered by WayVNC is supported by AVNC. That's why you see  "Authentication failed" error.

### Ideal Solution
AVNC should support the security types offered by WayVNC. That's what I am hoping to implement.

### Current Solution
1. First enable macOS security type in WayVNC:  https://github.com/gujjwal00/avnc/issues/249#issuecomment-2336399110
2. Then force AVNC to use macOS security type. In AVNC, long press on a server, select Edit, click on Advanced and change `Security` to `macOS`.
3. You should be able to connect now.  If for some reason you can't, post the logs from AVNC Settings => Tools => Logs.

---

> I can't try the test APK as the link gives a 404 error.
 
You have to be logged-in to download build artifacts. But macOS security type is already available in stable releases.

**Comment by Raptor-TK** [comment]
Thanks for your reply. Your clarification is good because it summarizes how to solve the issue in wayvnc.

However, in my case the vnc server in pikvm is not wayvnc. 
It is an ip based kvm for raspberry pi.
https://pikvm.org/

It's FAQ (https://docs.pikvm.org/vnc/#enabling-vnc-on-the-pikvm-side) describes the vnc server as:
Regular VNC/RDP/TeamViewer/Etc: OS -> Remote access server -> Network -> Client 
PiKVM: OS -> Video card PiKVM video capture -> PiKVM server -> Network -> Client

So pikvm is it's own vnc server.

It seems that bVNC Client works and apparently the pikvm author worked with bVNC to get it working.

As the errors when trying to connect to pi kvm were giving similar error as the wayvnc logs described in this issue and the title wasn't specific to wayvnc, I decided to post here. If it should go in a separate issue,  please let me know.

In pikvm there is not much option to change authentication to macos authentication. I tried the macOS option in the latest build of AVNC from fdroid and unfortunately get a different error when connecting to pikvm.

**Logs with macOS authentication set in AVNC:**
10-31 01:36:46.118  2438  3788 I NativeVnc: We have 1 security types to read
10-31 01:36:46.118  2438  3788 I NativeVnc: 0) Received security type 19
10-31 01:36:46.118  2438  3788 E NativeVnc: Unknown authentication scheme from VNC server: 19
10-31 01:36:46.119  2438  3788 E VncViewModel: Connection failed
10-31 01:36:46.119  2438  3788 E VncViewModel: java.io.IOException: Authentication failed
10-31 01:36:46.119  2438  3788 E VncViewModel: 	at com.gaurav.avnc.vnc.VncClient.connect(VncClient.kt:116)

**Logs when set to Auto authentication in AVNC:**
10-30 18:03:14.235 26986 29448 I NativeVnc: We have 1 security types to read
10-30 18:03:14.235 26986 29448 I NativeVnc: 0) Received security type 19
10-30 18:03:14.235 26986 29448 I NativeVnc: Selecting security type 19 (0/1 in the list)
10-30 18:03:14.235 26986 29448 I NativeVnc: Selected Security Scheme 19
10-30 18:03:14.326 26986 29448 I NativeVnc: Got VeNCrypt version 0.2 from server.
...
10-30 18:03:14.430 26986 29448 I NativeVnc: We have 3 security types to read
10-30 18:03:14.430 26986 29448 I NativeVnc: 0) Received security type 256
10-30 18:03:14.431 26986 29448 I NativeVnc: 1) Received security type 262
10-30 18:03:14.431 26986 29448 I NativeVnc: 2) Received security type 259
10-30 18:03:14.431 26986 29448 I NativeVnc: Selecting security type 262
10-30 18:03:14.527 26986 29448 E NativeVnc: Unsupported credential type requested
10-30 18:03:14.528 26986 29448 I NativeVnc: Reading credential failed
10-30 18:03:14.528 26986 29448 E NativeVnc: errnoToStr: (11 Try again)
10-30 18:03:14.528 26986 29448 E VncViewModel: Connection failed

Looking through other issues I read that AVNC uses libvncserver.
I found an issue in libvncserver that supposingly addresses the authentication issue in pikvm and the pikvm author answered some questions about it.
https://github.com/LibVNC/libvncserver/issues/476

It talks about how pikvm does authentication,  but I'm not sure if it has any clues to getting it working in AVNC.

I found another github issue from bVNC that describes how bVNC got it working with pikvm.
https://github.com/iiordanov/remote-desktop-clients/issues/245

Note that in both linked issues, mdevaev is the pikvm creator.

I believe that pikvm mentioned that pikvm uses VeNCrypt/Plain authorization type 262.

I think the actual fix for bVNC to work with pikvm was implementing X509Plain VeNCrypt security type 262 in place of TLSPlain security type 259.

**PR that fixed bVNC pikvm support**.
https://github.com/pikvm/kvmd/pull/28

I'm not sure if the issue is the same in AVNC.

However reading the posts here earlier you mention that type 262 is not supported by AVNC...

I also found the issue in libvncserver for implementing some missing types for pikvm.
Apparently these have been completed.
https://github.com/LibVNC/libvncserver/issues/458

And the PR here: https://github.com/LibVNC/libvncserver/pull/529

I'm not sure if these linked issues and PRs will give you a clue as to why it isn't working on AVNC.
I am trying to understand how the issue can be resolved but lack the understanding. Maybe you might get a hint from what I have compiled. It took a lot of digging about the problem to write this up.

I hope that i have explained the issue well enough.

**Comment by gujjwal00** [comment]
Thanks for the details, that's very helpful. Unfortunately root cause is the same, i.e. security type 262 not being supported in AVNC.

In the meantime, you should be able to enable VNCAuth in PiKVM (beware of the security risk). On the same FAQ page (https://docs.pikvm.org/vnc/#enabling-vnc-on-the-pikvm-side), follow the section `Optional for non-TigerVNC and NOT RECOMMENDED: Enable VNCAuth method`. If you enable that, remember to set `Security` in AVNC to `VNCAuth`.

**Comment by gujjwal00** [comment]
Just to give some details about security type 262 (more specifically, x509 sub-types of  Vencrypt):

This security type works similar to how browsers secure your communication with web servers. Web admins install a x509 certificate on web server, which presents the certificates to web browsers, and browsers verify that certificate using trusted root certificates.

From what I have seen on VNC side, typically a self-signed certificate is used with VNC server, and VNC clients employ trust-on-first-use, similar to how SSH works. That's what I indent to implement, but it will take some time.

**Comment by Raptor-TK** [comment]
This workaround is working with AVNC!

For others with the same issue, I have the following set in PiKVM's /etc/kvmd/override.yaml:
vnc:
    auth:
        vncauth:
            enabled: true
        vencrypt:
            enabled: false
    server:
        tls:
            ciphers: ""

This works with AVNC set to VNCAuth.

Thank you for your help!

**Comment by gujjwal00** [comment]
This will be available in next release. 
Test APK, if anybody wants to try it: https://github.com/gujjwal00/avnc/actions/runs/11652916670/artifacts/2138156348

This APK should work out-of the box without any changes on server or in AVNC.

**Comment by Raptor-TK** [comment]
Tested the APK and indeed it works out of the box for PiKVM.
It prompted to generate a certificate and then it connected fine.

**Comment by Kiisu-Master** [comment]
Works with wayvnc too, but there is a problem with automatic resize. Sometimes i will get a black screen or disconnects, especially when rotating the phone. It works again if i switch to MacOS security. Maybe a bug on wayvnc side but i don't know. It's a bit random.

**Comment by gujjwal00** [comment]
This is now available in v2.7.0
